cloudstack_plugin/virtual_machine.py

```diff
@@ -602,6 +606,62 @@ def disconnect_floating_ip(ctx, **kwargs):
     return True
 
 
+@operation
+def attach_volume(ctx, **kwargs):
+    """ Attach a volume to virtual machine.
+    """
+
+    cloud_driver = get_cloud_driver(ctx)
+
+    server_id = ctx.source.instance.runtime_properties[CLOUDSTACK_ID_PROPERTY]
+    volume_id = ctx.target.instance.runtime_properties[CLOUDSTACK_ID_PROPERTY]
+
+    server = get_vm_by_id(ctx, cloud_driver, server_id)
+    volume = get_volume_by_id(cloud_driver, volume_id)
+
+    if server is None:
+        raise NonRecoverableError('Server with id {0} not found'
+                                  .format(server_id))
+
+    if volume is None:
+        raise NonRecoverableError('Volume with id {0} not found'
+                                  .format(volume_id))
+
+    cloud_driver.attach_volume(node=server,
+                               volume=volume)
+
+
+@operation
+def detach_volume(ctx, **kwargs):
+    """ Detaches a volume and delete if expunge is requested.
+    """
+
+    cloud_driver = get_cloud_driver(ctx)
+
+    volume_id = ctx.target.instance.runtime_properties[CLOUDSTACK_ID_PROPERTY]
+    volume = get_volume_by_id(cloud_driver, volume_id)
+
+    if volume is None:
+        raise NonRecoverableError('Volume with id {0} not found'
+                                  .format(volume_id))
+
+    # Detach the volume
+    try:
+        cloud_driver.detach_volume(volume=volume)
+    except Exception as e:
+        ctx.logger.warn('Volume {0} may not have been detached: {1}'
+                        .format(volume, str(e)))
+        pass
+
+    # Expunge the volume if true in blueprint configuration
```

Comment:
why is this part needed here? its already implemented in delete...
detach should just detach, not delete